<a href="https://colab.research.google.com/github/martasaparicio/lematecX/blob/main/4.1-Diabetes.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Solved Exercise

## Introduction

In this exercise, we intend to predict whether a given patient is diabetic or not. To do this, we will take into account several characteristics of these patients, such as their age, blood pressure and body mass index. 

## Data

*   Number of observations = 768
*   Number of independent variables = 8
  1.   Pregnancies = number of pregnancies 
  1.   Glucose = glucose level (mg/dL)
  1.   BloodPressure = diastolic blood pressure (mmHg)
  2.   SkinThickness = triceps skinfold thickness (mm)
  2.   Insulin = insulin level (mu U/ml)
  2.   BMI = body mass index (kg/m^2)
  1.   DiabetesPedigreeFunction = genetic predisposition for diabetes (according to a specific function)
  2.   Age = age (years)
*   Number of dependent variables = 1
  1.   Outcome = tells us if a patient is diabetic (1) or not (0)

Data availabe at:  https://raw.githubusercontent.com/pmarcelino/datasets/master/diabetes.csv

**Note:** The dependent and independent variables are defined specifically using the information in the exercise's introduction. If in the introduction we were asked, for example, to predict the insulin level using the other variables, the dependent variable would be 'Insulin' and not 'Outcome'.

## Example

The first observation in the dataset that we will see in the Solution - the first row in the table that appears after doing `df` - refers to a patient who:

*   Has had `6` pregnancies
*   Has a glucose level of `148` mg/dL
*   Has a diastolic blood pressure of `72` mmHg 
*   Has a triceps skinfold thickness of `35` mm 
*   Has an insulin level of `0` mu U/ml 
*   Has a body mass index of `33.6` kg/m^2
*   Has a genetic predisposition for diabetes of `0.627` 
*   Is `50` years old
*   Is `diabetic`




# Solution

1.   Prepare data
1.   Explore data
2.   Train the model
2.   Evaluate the model

In [ ]:
# Import libraries
import pandas as pd  
import seaborn as sns

from sklearn.ensemble import RandomForestClassifier  
from sklearn.model_selection import cross_val_score

**Note:** If `FutureWarning` appears while you are importing the libraries, do not worry. This message appears associated with a library and serves only to inform you that some changes are ocurring in that library.

## 1. Prepare data

1.   Import data
2.   Remove observations with missing data

In [ ]:
# Import data
url = "https://raw.githubusercontent.com/pmarcelino/datasets/master/diabetes.csv"
df = pd.read_csv(url)
df

Pregnancies  Glucose  ...  Age  Outcome
0              6      148  ...   50        1
1              1       85  ...   31        0
2              8      183  ...   32        1
3              1       89  ...   21        0
4              0      137  ...   33        1
..           ...      ...  ...  ...      ...
763           10      101  ...   63        0
764            2      122  ...   27        0
765            5      121  ...   30        0
766            1      126  ...   47        1
767            1       93  ...   23        0

[768 rows x 9 columns]

*   The import was successful


In [ ]:
# Check for missing data
pd.isnull(df).sum()

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64

*   There is no data missing

## 2. Explore data

1.   Detect errors and anomalies

In [ ]:
# Check for errors or anomalies
df.describe()

Pregnancies     Glucose  ...         Age     Outcome
count   768.000000  768.000000  ...  768.000000  768.000000
mean      3.845052  120.894531  ...   33.240885    0.348958
std       3.369578   31.972618  ...   11.760232    0.476951
min       0.000000    0.000000  ...   21.000000    0.000000
25%       1.000000   99.000000  ...   24.000000    0.000000
50%       3.000000  117.000000  ...   29.000000    0.000000
75%       6.000000  140.250000  ...   41.000000    1.000000
max      17.000000  199.000000  ...   81.000000    1.000000

[8 rows x 9 columns]

*   There are errors and anomalies
*   The following variables are not expected to have a minimum value of 0: 
  *   Glucose level
  *   Diastolic blood pressure
  *   Triceps skinfold thickness
  *   Insulin level
  *   Body mass index

In [ ]:
# Remove observations with erros and anomalies
df.drop(df[df['Glucose']==0].index, inplace=True)
df.drop(df[df['BloodPressure']==0].index, inplace=True)
df.drop(df[df['SkinThickness']==0].index, inplace=True)
df.drop(df[df['Insulin']==0].index, inplace=True)
df.drop(df[df['BMI']==0].index, inplace=True)

In [ ]:
# Check for errors or anomalies
df.describe()

Pregnancies     Glucose  ...         Age     Outcome
count   392.000000  392.000000  ...  392.000000  392.000000
mean      3.301020  122.627551  ...   30.864796    0.331633
std       3.211424   30.860781  ...   10.200777    0.471401
min       0.000000   56.000000  ...   21.000000    0.000000
25%       1.000000   99.000000  ...   23.000000    0.000000
50%       2.000000  119.000000  ...   27.000000    0.000000
75%       5.000000  143.000000  ...   36.000000    1.000000
max      17.000000  198.000000  ...   81.000000    1.000000

[8 rows x 9 columns]

*   The observations previously identified as containing errors or anomalies have disappeared
*   The number of observations has gone from 768 to 392 (see `count`)

**Conclusions drawn from the data exploration**

*   There were obvious errors and anomalies and these were rectified

## 3. Train the model

In [ ]:
# Define the independent and dependent variables
X = df.drop('Outcome', axis=1)
y = df['Outcome']

In [ ]:
# Define the algorithm of the model
model = RandomForestClassifier(random_state=1143)

*   At a certain point in the Random Forest algorithm, there is a random division of the data
*   In order for this division to always be equal, we define the `random_state` argument
*   That way, if anyone were to run code that is exactly the same as ours, they would get the same results
*   We used the number 1143, but we could have used any other number

## 4. Evaluate the model

In [ ]:
# Evaluate the model using cross-validation (using the success rate)
score = cross_val_score(model, X, y, cv=5)
score

array([0.79746835, 0.67088608, 0.75641026, 0.80769231, 0.83333333])

*   We used five iterations (`cv=5`) and that is why we have a `score` with five values (one for each iteration) 
*   Scikit-learn uses the evaluation metric predefined in the model. For RandomForestClassifier, the default evaluation metric is success rate 
*   Since this is a classification problem, scikit-learn automatically uses stratified cross-validation


In [ ]:
# Estimate the model's performance
score_mean = score.mean()
score_mean

0.7731580655631288

*   Our model has an average success rate of 77% (approximately) 



In [ ]:
# Evaluate the model using cross-validation (using sensitivity)
score = cross_val_score(model, X, y, scoring='recall', cv=5)
score

array([0.61538462, 0.46153846, 0.42307692, 0.65384615, 0.73076923])

*   If we consider diabetes to be a serious disease and we want to assess the model's ability to identify all positive cases, we can use another evaluation metric
*   In this context, an appropriate evaluation metric would be sensitivity 
*   To use this metric in cross-validation, which can be found in the scikit-learn library, we have to pass the `scoring='recall'` argument in the `cross_val_score` function 



In [ ]:
# Estimate the model's performance
score_mean = score.mean()
score_mean

0.576923076923077

*   Our model has a sensitivity of 58% (approximately)
*   Since the maximum would be 100% - which would mean that, on average, our model would detect all positive cases - we find that, taking into account the sensitivity, the performance of the model is low